File: src/rating_bot/fsm.py
"""Conversation state that survives a restart but not an absent user.

aiogram's bundled storages are memory (lost on deploy) or Redis (not here), so
this is a small sqlite one. State older than ``ttl`` reads as absent, which is
what stops a forgotten /announce from swallowing an unrelated message next week.
"""

import json
import sqlite3
import time
from collections.abc import Mapping
from contextlib import contextmanager
from pathlib import Path
from typing import Any

from aiogram.fsm.state import State
from aiogram.fsm.storage.base import BaseStorage, StorageKey

IDLE_TTL_SECONDS = 300.0


class SqliteStorage(BaseStorage):
    def __init__(self, path: Path, ttl: float = IDLE_TTL_SECONDS):
        self.path = path
        self.ttl = ttl

    @contextmanager
    def _cursor(self):
        conn = sqlite3.connect(self.path)
        try:
            yield conn.cursor()
            conn.commit()
        finally:
            conn.close()

    @staticmethod
    def _key(key: StorageKey) -> tuple:
        return (
            key.bot_id,
            key.chat_id,
            key.user_id,
            key.thread_id or 0,
            key.destiny,
        )

    def _read(self, key: StorageKey) -> tuple[str | None, dict]:
        with self._cursor() as cur:
            rows = cur.execute(
                """select state, data, updated_at from fsm_state
                   where bot_id=? and chat_id=? and user_id=? and thread_id=? and destiny=?""",
                self._key(key),
            ).fetchall()
            if not rows:
                return None, {}
            state, data, updated_at = rows[0]
            if time.time() - updated_at > self.ttl:
                cur.execute(
                    """delete from fsm_state
                       where bot_id=? and chat_id=? and user_id=? and thread_id=? and destiny=?""",
                    self._key(key),
                )
                return None, {}
        return state, json.loads(data)

    def _write(self, key: StorageKey, state: str | None, data: dict) -> None:
        with self._cursor() as cur:
            cur.execute(
                """insert into fsm_state
                   (bot_id, chat_id, user_id, thread_id, destiny, state, data, updated_at)
                   values (?,?,?,?,?,?,?,?)
                   on conflict(bot_id, chat_id, user_id, thread_id, destiny) do update set
                   state = excluded.state, data = excluded.data, updated_at = excluded.updated_at""",
                (*self._key(key), state, json.dumps(data), time.time()),
            )

    async def set_state(self, key: StorageKey, state: State | str | None = None) -> None:
        resolved = state.state if isinstance(state, State) else state
        _, data = self._read(key)
        if resolved is None and not data:
            with self._cursor() as cur:
                cur.execute(
                    """delete from fsm_state
                       where bot_id=? and chat_id=? and user_id=? and thread_id=? and destiny=?""",
                    self._key(key),
                )
            return
        self._write(key, resolved, data)

    async def get_state(self, key: StorageKey) -> str | None:
        return self._read(key)[0]

    async def set_data(self, key: StorageKey, data: Mapping[str, Any]) -> None:
        state, _ = self._read(key)
        self._write(key, state, dict(data))

    async def get_data(self, key: StorageKey) -> dict[str, Any]:
        return self._read(key)[1]

    async def update_data(
        self, key: StorageKey, data: Mapping[str, Any]
    ) -> dict[str, Any]:
        state, current = self._read(key)
        current.update(data)
        self._write(key, state, current)
        return current

    async def get_value(
        self, storage_key: StorageKey, dict_key: str, default: Any | None = None
    ) -> Any | None:
        return self._read(storage_key)[1].get(dict_key, default)

    async def close(self) -> None:
        pass

File: internal/store/fsm.go
package store

import (
	"database/sql"
	"encoding/json"
	"time"
)

// IdleTTL is how long a half-finished conversation stays alive. Past it the
// state reads as absent, which is what stops a forgotten /announce from
// swallowing an unrelated message next week.
const IdleTTL = 300 * time.Second

// Key identifies one conversation. The shape is the Python bot's, so an
// in-flight conversation survives the switch to this binary.
type Key struct {
	BotID    int64
	ChatID   int64
	UserID   int64
	ThreadID int64
	Destiny  string
}

// State is a conversation's position plus whatever the flow collected so far.
type State struct {
	Name string
	Data map[string]any
}

func (s State) String(key string) string {
	value, _ := s.Data[key].(string)
	return value
}

// Numbers come back from JSON as float64; every number these flows store is a
// day count that fits an int exactly.
func (s State) Int(key string) int {
	value, _ := s.Data[key].(float64)
	return int(value)
}

const fsmWhere = `bot_id=? and chat_id=? and user_id=? and thread_id=? and destiny=?`

func (k Key) args() []any {
	return []any{k.BotID, k.ChatID, k.UserID, k.ThreadID, k.Destiny}
}

type FSM struct {
	db  *DB
	ttl time.Duration
}

func NewFSM(db *DB, ttl time.Duration) *FSM { return &FSM{db: db, ttl: ttl} }

func (f *FSM) Get(key Key) State {
	var (
		name      sql.NullString
		data      []byte
		updatedAt float64
	)
	err := f.db.conn.QueryRow(
		`select state, data, updated_at from fsm_state where `+fsmWhere, key.args()...,
	).Scan(&name, &data, &updatedAt)
	if err != nil {
		return State{Data: map[string]any{}}
	}
	if time.Since(time.Unix(0, int64(updatedAt*float64(time.Second)))) > f.ttl {
		f.Clear(key)
		return State{Data: map[string]any{}}
	}
	state := State{Name: name.String, Data: map[string]any{}}
	if err := json.Unmarshal(data, &state.Data); err != nil {
		state.Data = map[string]any{}
	}
	return state
}

func (f *FSM) Set(key Key, state State) error {
	if state.Name == "" && len(state.Data) == 0 {
		return f.Clear(key)
	}
	data, err := json.Marshal(state.Data)
	if err != nil {
		return err
	}
	var name any
	if state.Name != "" {
		name = state.Name
	}
	args := append(key.args(), name, data, float64(time.Now().UnixNano())/float64(time.Second))
	_, err = f.db.conn.Exec(
		`insert into fsm_state
		 (bot_id, chat_id, user_id, thread_id, destiny, state, data, updated_at)
		 values (?,?,?,?,?,?,?,?)
		 on conflict(bot_id, chat_id, user_id, thread_id, destiny) do update set
		 state = excluded.state, data = excluded.data, updated_at = excluded.updated_at`,
		args...,
	)
	return err
}

// SetName moves the conversation to a new step, keeping what it has collected.
func (f *FSM) SetName(key Key, name string) error {
	state := f.Get(key)
	state.Name = name
	return f.Set(key, state)
}

func (f *FSM) Update(key Key, values map[string]any) error {
	state := f.Get(key)
	for k, v := range values {
		state.Data[k] = v
	}
	return f.Set(key, state)
}

func (f *FSM) Clear(key Key) error {
	_, err := f.db.conn.Exec(`delete from fsm_state where `+fsmWhere, key.args()...)
	return err
}
